### PR TITLE
#1649 OrderPOCreate process patch for bug 

### PR DESCRIPTION
**Pull request by homebeaver**
Hi @yamelsenih , @e-Evolution ,

this is a replacement for PR https://github.com/adempiere/adempiere/pull/1651 

It fixes bug https://github.com/adempiere/adempiere/issues/1649

### CODE REVIEW

File: base/src/org/compiere/process/OrderPOCreate.java [hunk 1]
```diff
@@ -38,54 +44,50 @@
...
-	/**	Vendor				*/
-	private int			p_Vendor_ID;
-	/**	Sales Order			*/
-	private int			p_C_Order_ID;
-	/** Drop Ship			*/
-	private boolean		p_IsDropShip = false;
 	
+	private int m_explosion_level = 0;
+	private Set<MPPProductBOMLine> bomSet = new HashSet<MPPProductBOMLine>(); 
+	private int explosion(MPPProductBOM bom) throws AdempiereUserError {
```
Comment (yamelsenih) [inline base/src/org/compiere/process/OrderPOCreate.java]: Hi @homebeaver i don't know what is the reason for add BOM process here, BOM process is created when you complete a Sales Order, see it here: https://github.com/adempiere/adempiere/blob/develop/base/src/org/compiere/model/MOrder.java#L1239

All changes is ok but in don't agree with add a separed process for make **exprosion** is it is part of business model for MOrder

Best regard

Comment (homebeaver) [reply base/src/org/compiere/process/OrderPOCreate.java]: @yamelsenih , let me explain:

if you have a BOM-product in your sales order and you create the purchase order from it, so you cannot purchase the BOM-product itsself, You have to purchase all the components and produce the product (make-to-order). This was not considered in `OrderPOCreate` .
I did not implement purchase components, nor make-to-order. This is heavy work. 
But at least I send a warning:
```
   // TODO purchase component and prepare BOM production
  log.warning(mProduct.toString() + " is BOM product, will not be purchased (not implemented)");
```
It's up to you (or us) to decide what is better
- send a warning on not implemeted feature
- do nothing , remove the explosion and do not inform the user about the hole

This have some interfaces to production.
What do you think? What is the opinion of Mario @marcalwestf  ?
regards

**Comment by marcalwestf** [comment]
giving it a look....

**Comment by marcalwestf** [comment]
hi Eugen,
I agree with you on the status issue.

Here my comments about BOM explosion + PO.

You wrote
> if you have a BOM-product in your sales order and you create the purchase order from it, so you cannot purchase the BOM-product itsself

My comments: we have here to remember that it is also possible that you can buy a product with a BOM. This would be the case when you mark the product as _isPurchased_ and it has also a BOM. This is not uncommon.

Then you stated
> You have to purchase all the components and produce the product (make-to-order). 

My comments: the correct procedure would be to create first a _PPOrder_ (Libero) or a _M_ProductionBatch_ (Manufacturing Light). This two documents generate demand, which would be satisfied with Purchase Orders or Productions. 

Also, different demands are usually **bundled** into a single Purchase Order for example using MRP or Replenishment reports. It is uncommon to directly satisfy a demand for a single Sales Order. This would disrupt the practice implemented.

Thus IMHO I would agree **not to include the BOM explosion and Purchase Order for the exploded products**.

I'd suggest to leave only the changes concerning the status and not include the explosion.

File: base/src/org/compiere/process/OrderPOCreate.java (hunk 1 above)
Comment (marcalwestf) [reply base/src/org/compiere/process/OrderPOCreate.java]: see my answer below

**Comment by homebeaver** [comment]
OK. Thanks Mario, @marcalwestf for the explanation

**Comment by marcalwestf** [comment]
Eugen, would you do the suggested changes? I'd accept the pull request then.
(it's a pity because of the nice and complex code)

**Comment by homebeaver** [comment]
Mario @marcalwestf I am on holidays for the next 3 weeks.
Can you or @yamelsenih please delete or comment out the unnecessary lines in `OrderPOCreate.java` tkx+regards

**Comment by homebeaver** [comment]
Hi Mario, you 've looked to a wrong branch. The correct is
https://github.com/klst-de/adempiere/commits/develop+1649 (with a plus) Try tihis 👍 
regards E.

**Comment by marcalwestf** [comment]
hi Eugen @homebeaver , I looked again at the pull request.

I compared the changes and they look OK.
I like this PR because you have also improved the code and updated the process.

Only with the following snippet I have some doubts:
```
if(purchaseProduct && mProduct.isStocked()) { // check whether the product is onHand
  MStorage[] mStorages = MStorage.getOfProduct(getCtx(), M_Product_ID, get_TrxName());
  for(int s = 0; s < mStorages.length; s++) {
	log.config("mStorage="+mStorages[s]);
	if(mStorages[s].getQtyOnHand().compareTo(soLines[i].getQtyOrdered()) > -1 ) {
		log.warning(mProduct.toString() + " is onHand, will not be purchased");
		purchaseProduct = false;
	}
  }
}
```
1.- This would conflict with backward compatibility: up to now, the process worked **without** warehouse check. What happens with existing implementations that rely on the fact that the Purchase Order is generated "as a whole" out of the Sales Order?  
With the the pull request as it it, this would not be possible because of the _quantity on hand_ check.
This could be overcome with an additional parameter in the process.

2.- When checking storage (for example using the parameter suggested above), then you should first add _quantity on hand_ for all warehouses and only then compare, instead of just comparing just one warehouse:
```
//	Qty On Hand
BigDecimal qtyOnHand = Env.ZERO;
MStorage[] storages = MStorage.getWarehouse (getCtx(), 0,
    mProduct.getM_Product_ID(), soLines[i].getM_AttributeSetInstance_ID(),
	null, MClient.MMPOLICY_FiFo.equals(MMPolicy), true, soLines[i].getM_Locator_ID(), get_TrxName());

if(storages != null
		&& storages.length > 0) {
	for(MStorage storage : storages) {
		qtyOnHand = qtyOnHand.add(storage.getQtyOnHand());
	}
}

if(qtyOnHand.compareTo(soLines[i].getQtyOrdered()) > -1 ) {
	log.warning(mProduct.toString() + " is onHand, will not be purchased");
	purchaseProduct = false;
	}
```

If you think this is too much a trouble, you can comment out the snippets lines and we can discuss this matter of checking _on hand quantity_ later.

Sorry to bother you again, but I think backward compatibility is an important matter.
Thanks for your understanding.

**Comment by marcalwestf** [comment]
hi Eugen @homebeaver as we're accepting the last pull requests prior to the 3.9.1 release, I wonder if you can do something regarding this pull request?
Thanks

**Comment by homebeaver** [comment]
> 
> 
@marcalwestf wrote
> ... I wonder if you can do something regarding this pull request?
Hi Mario,
let us switch it after 3.9.1 - no hasty reaction at finish line.
PS: and thanks for the comments
bye EUGen

**Comment by yamelsenih** [comment]
Replaced by https://github.com/adempiere/adempiere/pull/2201

Thanks @homebeaver

**Pull request closed without merging**